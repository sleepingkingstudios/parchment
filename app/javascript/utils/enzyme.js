import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
