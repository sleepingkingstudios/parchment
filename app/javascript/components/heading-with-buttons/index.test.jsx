import React from 'react';
import { shallow } from 'enzyme';

import HeadingWithButtons from './index';

describe('<HeadingWithButtons />', () => {
  const buttonData = {
    label: 'Button Label',
    url: 'path/to/url',
  };
  const children = 'Heading Label';
  const defaultProps = { buttons: [buttonData], children };

  it('should wrap the contents in a div', () => {
    const rendered = shallow(<HeadingWithButtons {...defaultProps} />);

    expect(rendered).toHaveDisplayName('div');
    expect(rendered).toHaveClassName('heading-with-buttons');
  });

  it('should render the heading', () => {
    const rendered = shallow(<HeadingWithButtons {...defaultProps} />);
    const heading = rendered.find('h1');

    expect(heading).toExist();
    expect(heading).toHaveProp({ children });
  });

  it('should render the button', () => {
    const rendered = shallow(<HeadingWithButtons {...defaultProps} />);
    const button = rendered.find('LinkButton');
    const { label, url } = buttonData;

    expect(button).toExist();
    expect(button).toHaveProp('children', label);
    expect(button).toHaveProp({ url });
  });

  it('should match the snapshot', () => {
    const rendered = shallow(<HeadingWithButtons {...defaultProps} />);

    expect(rendered).toMatchSnapshot();
  });

  describe('with a fancy button', () => {
    const fancyButtonData = {
      buttonStyle: 'danger',
      label: 'Self Destruct',
      outline: true,
      url: 'path/to/url',
    };
    const props = { ...defaultProps, buttons: [fancyButtonData] };

    it('should render the button', () => {
      const rendered = shallow(<HeadingWithButtons {...props} />);
      const button = rendered.find('LinkButton');
      const {
        buttonStyle,
        label,
        outline,
        url,
      } = fancyButtonData;

      expect(button).toExist();
      expect(button).toHaveProp('children', label);
      expect(button).toHaveProp({ buttonStyle });
      expect(button).toHaveProp({ outline });
      expect(button).toHaveProp({ url });
    });
  });

  describe('with many buttons', () => {
    const multiButtonData = [
      {
        label: 'First Button',
        url: 'button/one',
      },
      {
        label: 'Second Button',
        url: 'button/two',
      },
      {
        label: 'Third Button',
        url: 'button/three',
      },
    ];
    const props = { ...defaultProps, buttons: multiButtonData };

    it('should render the buttons in reverse order', () => {
      const rendered = shallow(<HeadingWithButtons {...props} />);
      const buttons = rendered.find('LinkButton');

      expect(buttons.length).toEqual(3);

      buttons.forEach((button, index) => {
        const { label, url } = multiButtonData[2 - index];

        expect(button).toHaveProp('children', label);
        expect(button).toHaveProp({ url });
      });
    });

    it('should match the snapshot', () => {
      const rendered = shallow(<HeadingWithButtons {...props} />);

      expect(rendered).toMatchSnapshot();
    });
  });
});
