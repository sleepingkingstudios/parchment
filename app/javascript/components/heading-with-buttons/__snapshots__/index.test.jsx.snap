// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeadingWithButtons /> should match the snapshot 1`] = `
<div
  className="heading-with-buttons"
>
  <h1>
    Heading Label
  </h1>
  <LinkButton
    block={false}
    buttonSize={null}
    buttonStyle="primary"
    className="btn-sm float-right"
    key="button-0"
    link={false}
    outline={false}
    url="path/to/url"
  >
    Button Label
  </LinkButton>
</div>
`;

exports[`<HeadingWithButtons /> with many buttons should match the snapshot 1`] = `
<div
  className="heading-with-buttons"
>
  <h1>
    Heading Label
  </h1>
  <LinkButton
    block={false}
    buttonSize={null}
    buttonStyle="primary"
    className="btn-sm float-right"
    key="button-0"
    link={false}
    outline={false}
    url="button/three"
  >
    Third Button
  </LinkButton>
  <LinkButton
    block={false}
    buttonSize={null}
    buttonStyle="primary"
    className="btn-sm float-right"
    key="button-1"
    link={false}
    outline={false}
    url="button/two"
  >
    Second Button
  </LinkButton>
  <LinkButton
    block={false}
    buttonSize={null}
    buttonStyle="primary"
    className="btn-sm float-right"
    key="button-2"
    link={false}
    outline={false}
    url="button/one"
  >
    First Button
  </LinkButton>
</div>
`;
